pkg/auth/db_service.go

```diff
@@ -867,3 +882,304 @@ func (s *DBAuthService) markTokenSingleUse(ctx context.Context, tokenID string,
+			return nil, err
+		}
+		kvUser := &model.User{
+			ID:       uuid.New().String(),
+			BaseUser: dbUser.BaseUser,
+		}
+		key := model.KVUserPath(dbUser.Username)
+		value, err := proto.Marshal(model.ProtoFromUser(kvUser))
+		if err != nil {
+			return nil, err
+		}
+
+		if err = je.Encode(kv.Entry{
+			PartitionKey: []byte(model.PartitionKey),
+			Key:          []byte(key),
+			Value:        value,
+		}); err != nil {
+			return nil, err
+		}
+
+		userID2Details[dbUser.ID] = userDetails{
+			kvID: kvUser.ID,
+			name: kvUser.Username,
+		}
+	}
+	return userID2Details, nil
+}
+
+func exportGroups(ctx context.Context, d *pgxpool.Pool, je *json.Encoder) (IDToName, error) {
+	rows, err := d.Query(ctx, "SELECT * FROM auth_groups ORDER BY created_at ASC")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	groupID2Name := make(IDToName)
+	scanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		dbGroup := model.DBGroup{}
+		err = scanner.Scan(&dbGroup)
+		if err != nil {
+			return nil, err
+		}
+		kvGroup := &model.Group{
+			ID:        uuid.New().String(),
+			BaseGroup: dbGroup.BaseGroup,
+		}
+		model.ConvertGroup(&dbGroup)
+		key := model.KVGroupPath(dbGroup.DisplayName)
+		value, err := proto.Marshal(model.ProtoFromGroup(kvGroup))
+		if err != nil {
+			return nil, err
+		}
+		if err = je.Encode(kv.Entry{
+			PartitionKey: []byte(model.PartitionKey),
+			Key:          []byte(key),
+			Value:        value,
+		}); err != nil {
+			return nil, err
+		}
+	}
+	return groupID2Name, nil
+}
+
+func exportPolicies(ctx context.Context, d *pgxpool.Pool, je *json.Encoder) (IDToName, error) {
+	rows, err := d.Query(ctx, "SELECT * FROM auth_policies ORDER BY created_at ASC")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	policyID2Name := make(IDToName)
+	scanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		dbPolicy := model.DBPolicy{}
+		err = scanner.Scan(&dbPolicy)
+		if err != nil {
+			return nil, err
+		}
+		key := model.KVPolicyPath(dbPolicy.DisplayName)
+		value, err := proto.Marshal(model.ProtoFromPolicy(&dbPolicy.BasePolicy, strconv.Itoa(dbPolicy.ID)))
+		if err != nil {
+			return nil, err
+		}
+		if err = je.Encode(kv.Entry{
+			PartitionKey: []byte(model.PartitionKey),
+			Key:          []byte(key),
+			Value:        value,
+		}); err != nil {
+			return nil, err
+		}
+		policyID2Name[dbPolicy.ID] = dbPolicy.DisplayName
+	}
+	return policyID2Name, nil
+}
+
+func exportUserGroups(ctx context.Context, d *pgxpool.Pool, je *json.Encoder, usersDetails UserIDToDetails, groupsNames IDToName) error {
+	type relation struct {
+		UserID  int64 `db:"user_id"`
+		GroupID int   `db:"group_id"`
+	}
+	rows, err := d.Query(ctx, "SELECT * FROM auth_user_groups")
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+	scanner := pgxscan.NewRowScanner(rows)
+	for rows.Next() {
+		ug := relation{}
+		err = scanner.Scan(&ug)
+		if err != nil {
+			return err
+		}
+		username := usersDetails[ug.UserID].name
```

Comment:
Here and elsewhere: I think you should be careful when reading from all maps. Default values are not your friends and they will not return an error but spam the kv and probably end up in a non functional kv state.

Reply:
Excellent point. Fixed